### DISCUSSION TITLE
Downgrade functionality vs upgrade tests

### DISCUSSION THREAD

**Discussion by avikivity** [scylladb/scylladb#21698] [General]
I'm looking at refactoring [multi_column_range_accumulator](https://github.com/scylladb/scylladb/blob/cb6c55209aa365216293e4fe163c5fe21ca985d4/cql3/restrictions/statement_restrictions.cc#L2126-L2167) as part of a large series.

What this code does is take inputs such as x > 7, x <11, x IN (2, 9) and solve for x. This involves intersecting intervals (generated by x > ?, x < ?)  with each other and with the point solutions (to x = ? and x IN ?).

The problem is that this code is almost unused. Validation code rejects having more than one start bound or one end bound, and mixing equality, IN, and other relations with each other. This code only intersects the interval generated by the start bound and the interval generated by the end bound, if both are present.

If I just refactor it, the refactored code would be just as untested.

I could
 - refactor the code to only support what passes validation (losing some unused functionality)
 - expose the functionality only for tests, and test it, in anticipation of sometime relaxing the validation (but possibly doing useless work)
 - relax the validation now (would derail the refactoring)

@nyh what do you recommend here?

**Comment by avikivity** [comment]
I found a solution that doesn't require any of these bad tradeoffs.

**Comment by nyh** [comment]
I don't know what is a "discussion" but I see it's still open. What did you end up doing?

I general, I think you have two options:
1. Test via CQL,  after adding a "support experimental CQL" experimental option. In the cqlpy framework, I would use `config_value_context` to enable this option just for a single test (you wouldn't want to enable it for all tests - it might possibly break some).
2. Test with a unit test which calls your internal functions and checks their responses, not through CQL.

In the past you did a lot of option 2, but I personally prefer option 1. One of benefits of option 1 is that when you're done, you are left with a great set of **regression tests** - tests that will allow you to do further refactoring later without worrying that this feature break. Unit tests which check internal functions often need to be thrown away (or heavily rewritten) when you refactor those internal functions.
